src/audio/dai.c

```diff
@@ -423,6 +423,23 @@ static int dai_params(struct comp_dev *dev)
 		return -EINVAL;
 	}
 
+	/* set dma channel */
+	if (dd->chan == DMA_CHAN_INVALID)
+		/* get dma channel at first config only */
+		dd->chan = dma_channel_get(dd->dma, dev->params.stream_tag);
```

Comment:
Actually dev->params.stream_tag is for host dma, not link dma. And host dma channel is not always equal to link dma channel.

Reply:
No, this doesn't work, we should add a flag into params for link_stream_tag.

Reply:
@RanderWang @keyonjie for us link DMA channel is always equal to host DMA channel, isnt it? 
Thats the whole argument about why we're having all kinds of issues now. Care to explain your thoughts about a separate link_stream_tag? When would it be different from the host DMA tag? And if it is, how do you make sure that things dont break when you couple them foer setting the DMA format?

Reply:
And before you call out "this doesnt work", I'd appreciate if you could try it out and show me the "not working" usecase. Thanks!

Reply:
> And before you call out "this doesnt work", I'd appreciate if you could try it out and show me the "not working" usecase. Thanks!

Sorry, didn't read the whole commit message, now I have more time to write a longer comment, I mean we need solution for M:N in a one shot, and it's good to leave link DMA channel indices allocation/decision from driver side(here your solution is assuming fixing them to host indices and no any flexibility for future).

Can you consider suggestion like below to solve M:N mapping issues: 

```
++struct sof_hda_stream_tag {
++	uint32_t dai_index;	/**< index of this type dai */
++    uint32_t link_stream_tag;
++};

struct sof_ipc_stream_params {
...
	uint32_t rate;
--	uint16_t stream_tag;
++	uint16_t host_stream_tag;
	uint16_t channels;
...
++    uint16_t hda_stream_cnt;
++    struct sof_hda_stream_tag link_stream_tag[0];
} __attribute__((packed));
```
w.r.t. channel indices allocation in driver side, we allocate them in each PCM hw_params, taking 1:N mapping as example, that is, allocate i for host_stream_tag, and i, j, k, ... for the N link_stream_tag array, the couple(to i) --> decouple for stream format setting here is safe here.

OTOH, for M:1 mapping(we have corresponding M FE PCMs in driver side), the 1st (hw_params() running) of M PCM will be allocated to i for host_stream_tag and link_stream_tag, the hw_params() for remaining PCMs only allocate j, k, ... for host_stream_tag in driver side, and in FW side, the HDA dai will only accept the 1st hw_params and ignore the remaining ones. So here it is also safe.

Better to do one more thing in driver side that I talked to Rander, make sure we will couple the channel only at the case that both host channel and link channel are idle, this can improve our robustness for concurrent PCMs, and that is actually recommended from hardware docs.

Reply:
@keyonjie appreciate your thoughts but M:N is quite a complicated case and it cannot be fixed overnight. We do not have the ability in the FW to handle multiple BE's at this point. It needs a larger discussion. I cannot implement a half-baked solution with the hopes that someday it might be useful to fix something. With the current status of HDA platforms, we should resign to the fact that M:N is not supported and will be added later.

Reply:
@ranj063  for two FEs share one BE,   the dai_param would be called two times and one of them the tag should be different.

Reply:
And also for some stream with only BE like tone generator,  there is no host stream, right ?

Reply:
@RanderWang are you sure about this? Why would the tag be different? The link DMA ch is already in use and should not be changed when another FE is activated. In this case, the mixer would automatically come into play and it would mix the 2 streams. Why would you change the BE DAI channel?

Reply:
> And also for some stream with only BE like tone generator, there is no host stream, right ?

Agreed, but you would still need to hw_params in this case when you can pass in the stream_tag.

Reply:
> @keyonjie appreciate your thoughts but M:N is quite a complicated case and it cannot be fixed overnight. We do not have the ability in the FW to handle multiple BE's at this point. It needs a larger discussion. I cannot implement a half-baked solution with the hopes that someday it might be useful to fix something. With the current status of HDA platforms, we should resign to the fact that M:N is not supported and will be added later.

Hi @ranj063 sorry but have to request more than only 1:1 mapping works, at least we should provide the extendable possibility for future, otherwise, we have to go back and forth about it, that consume our effort but no actually improvement from user experience point of view.

Reply:
@keyonjie @RanderWang @ranj063 please make sure the programming flow and ABI is not restricted to 1:1 mappings as other HDA DMA HW may support 1:N or N:1.

Reply:
@lgirdwood i now have some insight into how to make the PR a bit more generic for the M:N case. I will have the updates by tomorrow.